### PR TITLE
Disallow re-entering wasm while off main stack

### PR DESCRIPTION
**Pull request by frank-emrich**
This PR addresses an issue discussed in #109:
Correctly handling the case where we re-enter wasm while already on a continuation stack is difficult. For the time being, we therefore disallow this. This PR adds the necessary logic to detect this.

Concretely, in `invoke_wasm_and_catch_traps`, we inspect the chain of nested wasm (+ host) invocations, represented by the linked list of `CallThreadState` objects maintained in `wasmtime_runtime:traphandlers`. To this end, for those `CallThreadState` objects that represent execution of wasm, we store a pointer to the corresponding `Store`'s `StackChainCell`.

Please note that the diff of the test file `typed_continuations.rs` looks unnecessarily scary: I moved the existing tests into a module `wasi`, and added modules `test_utils` and `host`.

### CODE REVIEW

File: crates/runtime/src/traphandlers.rs [hunk 1]
```diff
@@ -207,22 +208,31 @@ pub unsafe fn catch_traps<'a, F>(
...
-    let result = CallThreadState::new(signal_handler, capture_backtrace, capture_coredump, *limits)
-        .with(|cx| {
-            traphandlers::wasmtime_setjmp(
-                cx.jmp_buf.as_ptr(),
-                call_closure::<F>,
-                &mut closure as *mut F as *mut u8,
-                caller,
-            )
-        });
+    let callee_stack_cain = VMContext::try_from_opaque(callee)
```
Comment (dhil) [inline crates/runtime/src/traphandlers.rs:218]:
```suggestion
    let callee_stack_chain = VMContext::try_from_opaque(callee)
```

File: crates/runtime/src/traphandlers.rs [hunk 2]
```diff
@@ -207,22 +208,31 @@ pub unsafe fn catch_traps<'a, F>(
...
+        .map(|vmctx| Instance::from_vmctx(vmctx, |i| *i.stack_chain() as *const StackChainCell));
+
+    let result = CallThreadState::new(
+        signal_handler,
+        capture_backtrace,
+        capture_coredump,
+        *limits,
+        callee_stack_cain,
```
Comment (dhil) [inline crates/runtime/src/traphandlers.rs:226]:
```suggestion
        callee_stack_chain,
```

File: crates/runtime/src/traphandlers.rs (hunk 1 above)
Comment (dhil) [reply crates/runtime/src/traphandlers.rs:218]: If I understand this code correctly, then there is a new linear cost to enter Wasm? I suppose the cost may be negligible for now. Nonetheless, I wonder whether we can make this computation lazy? Do we always require immediate access to the `StackChainCell`?

**Review by dhil — COMMENTED** [review]
LGTM.

File: crates/runtime/src/traphandlers.rs (hunk 1 above)
Comment (dhil) [reply crates/runtime/src/traphandlers.rs]: I think it is worth documenting for our future selves.

Comment (frank-emrich) [reply crates/runtime/src/traphandlers.rs]: Okay, I've extended the existing documentation in `invoke_wasm_and_catch_traps`

**Pull request merged**